### PR TITLE
[Example] Add OBB external feature reclassifier

### PR DESCRIPTION
**Pull request by jackkowalik**
Adds a bolt-on classification example for OBB models using an external frozen spatial feature encoder. Includes a reproducible benchmark on DOTA v1.0 with a public evaluation key (no signup required). Script, README, and expected outputs included.

Discussed in #23821 at Glenn's suggestion.

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds a new end-to-end OBB example that benchmarks a bolt-on external feature reclassifier for improving YOLO OBB classification on DOTA without changing detector weights or detection outputs. 🚀

### 📊 Key Changes
- Introduces a new example folder: `examples/OBB-External-Feature-Reclassifier/`.
- Adds a detailed `README.md` with:
  - Method overview and workflow
  - DOTA v1.0 benchmark results for YOLOv8l-OBB and YOLO26l-OBB
  - Setup, dataset prep, usage commands, caching options, and expected output
- Adds `obb_feature_reclassifier.py` implementing:
  - DOTA label parsing and image tiling for large-scene OBB evaluation
  - YOLO OBB inference and prediction-to-GT matching
  - Optional YOLOv8 pre-NMS full-score extraction mode
  - External API-based crop encoding (batched with retry/fallback handling)
  - 5-fold scene-level cross-validation with LightGBM and class-wise reporting
- Adds `requirements.txt` for reproducible example dependencies.

### 🎯 Purpose & Impact
- Demonstrates a practical, non-invasive way to improve OBB classification by post-processing features rather than retraining detectors. 🧠
- Provides a reproducible benchmark pipeline users can run on DOTA and adapt to other domains.
- Expands Ultralytics examples with an OBB-focused, research-style integration pattern that may inspire downstream classification improvements.
- Highlights compatibility considerations between YOLOv8 score extraction and YOLO26 architecture behavior for fair evaluation.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jackkowalik, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help streamline review—an Ultralytics engineer will also assist you shortly. Please review the checklist below:
-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee
For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀🙂

File: examples/OBB-External-Feature-Reclassifier/obb_feature_reclassifier.py [hunk 1]
```diff
@@ -0,0 +1,630 @@
...
+from PIL import Image
+from sklearn.metrics import f1_score
+from sklearn.model_selection import GroupKFold
+from tqdm import tqdm
+from ultralytics import YOLO
+from ultralytics.data.augment import LetterBox
+
+API_URL = "https://api.authorize.earth/v1/spatial/encode"
+API_KEY = "sk_ae_1289e06722678304582fcde59ba5573aab73f32d91601d22022f2890ed4a7833"
+BATCH_SIZE = 64
```
Comment (UltralyticsAssistant) [inline examples/OBB-External-Feature-Reclassifier/obb_feature_reclassifier.py:123]: ⚠️ **HIGH**: A raw API key is hardcoded in source. Even if intended as a public eval key, embedding credentials directly in code is a security and abuse risk and makes rotation difficult. Load from environment/config and fail with a clear error if unset.

File: examples/OBB-External-Feature-Reclassifier/obb_feature_reclassifier.py [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+            crops = np.load(crop_path)
+            for i, r in enumerate(records):
+                r["crop"] = crops[i]
+            print(f"  {len(records)} detections loaded")
+        else:
+            model = YOLO(args.model)
+            records = run_detection(model, args.images, args.labels, full_scores=args.full_scores)
+            np.save(crop_path, np.array([r["crop"] for r in records]))
+            serializable = [
+                {k: v if not isinstance(v, np.ndarray) else v.tolist() for k, v in r.items() if k != "crop"}
```
Comment (UltralyticsAssistant) [inline examples/OBB-External-Feature-Reclassifier/obb_feature_reclassifier.py:664]: ⚠️ **HIGH**: `--skip-detection` only checks for `records.npz` before loading both records and crops, so a missing or stale `crops.npy` will raise at runtime. Gate this branch on both cache files (or handle missing crops gracefully).

**Suggested change:**
```suggestion
        if args.skip_detection and os.path.exists(rec_path) and os.path.exists(crop_path):
```

File: examples/OBB-External-Feature-Reclassifier/obb_feature_reclassifier.py [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+        else:
+            features = encode_all(records)
+            np.save(feat_path, np.array(features))
+
+        evaluate(records, features, Path(args.model).stem)
+
+    else:
+        parser.print_help()
+
+
```
Comment (UltralyticsAssistant) [inline examples/OBB-External-Feature-Reclassifier/obb_feature_reclassifier.py:685]: 💡 **MEDIUM**: Cached encoder features are loaded without validating they align with the current `records` length. If cache contents come from a different run/model/data slice, this can cause shape errors or silently evaluate mismatched features. Add a length check and recompute on mismatch.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Solid, well-structured example with clear pipeline stages and practical caching/retry flow. I found three important issues to address before merge: one definite cache-loading crash path, one cache-consistency validation gap for features, and one credential-handling security concern.

💬 Posted 3 inline comments

**Comment by jackkowalik** [comment]
I have read the CLA Document and I sign the CLA

File: examples/OBB-External-Feature-Reclassifier/obb_feature_reclassifier.py (hunk 1 above)
Comment (jackkowalik) [reply examples/OBB-External-Feature-Reclassifier/obb_feature_reclassifier.py:123]: Intentionally public eval key with per-IP rate limiting (3K/hr per IP, 15K/hr global). Designed to run with zero setup.

File: examples/OBB-External-Feature-Reclassifier/obb_feature_reclassifier.py (hunk 2 above)
Comment (jackkowalik) [reply examples/OBB-External-Feature-Reclassifier/obb_feature_reclassifier.py:664]: Fixed

File: examples/OBB-External-Feature-Reclassifier/obb_feature_reclassifier.py (hunk 3 above)
Comment (jackkowalik) [reply examples/OBB-External-Feature-Reclassifier/obb_feature_reclassifier.py:685]: Fixed

File: examples/OBB-External-Feature-Reclassifier/README.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+The pattern works with a wide variety of detectors, not just YOLO. Crop your detections, send them to the encoder, concatenate, and classify.
+
+## Setup
+
+```bash
+git clone https://github.com/ultralytics/ultralytics
+cd ultralytics/examples/OBB-External-Feature-Reclassifier
+pip install ultralytics==8.4.21 lightgbm scikit-learn numpy opencv-python requests pillow tqdm
+export AUTHORIZE_EARTH_API_KEY=sk_ae_1289e06722678304582fcde59ba5573aab73f32d91601d22022f2890ed4a7833
```
Comment (onuralpszr) [inline examples/OBB-External-Feature-Reclassifier/README.md]:
```suggestion
export AUTHORIZE_EARTH_API_KEY=your_earth_api_key
```


Revoke your API key I don't know it is fake or not but if  it is not, please revoke it.

**Review by onuralpszr — CHANGES_REQUESTED** [review]

File: examples/OBB-External-Feature-Reclassifier/README.md (hunk 4 above)
Comment (jackkowalik) [reply examples/OBB-External-Feature-Reclassifier/README.md]: Fixed

**Comment by jackkowalik** [comment]
Updated the script and README to clarify the cross-validation protocol. DOTA images are up to 20,000×20,000 px and must be tiled into 1024×1024 patches with 200px overlap. The default evaluation uses tile-level GroupKFold, where tiles from the same original image at different spatial positions can appear in different folds. This mirrors production deployment where labeled data from the operating region is available and new detections need classification. The LightGBM is trained once on labeled examples and then classifies all new detections at inference without retraining.

For strict image-level holdout (entire original images held out per fold), a --strict-scene-split flag is now available. Under strict splitting the DOTA improvement is marginal to negative, consistent with the encoder capturing spatial context that transfers within a deployment region. Other benchmarks with unambiguous scene boundaries (xView2 disaster events, RarePlanes satellite passes) show error reduction under strict scene-level splits.

**Comment by glenn-jocher** [comment]
Thanks for the clarification and for adding `--strict-scene-split`; that makes the evaluation assumptions much clearer, and keeping this as a self-contained example under [Ultralytics examples](https://github.com/ultralytics/ultralytics/tree/main/examples) with exact README commands plus expected outputs should make review much easier for an OBB workflow like the one described in the [OBB docs](https://docs.ultralytics.com/tasks/obb/).

**Comment by glenn-jocher** [comment]
Thanks for the follow-up updates; please keep this scoped to the single OBB example, make sure the README commands and any supporting tests stay aligned with the current behavior, and re-request review when ready.

**Comment by glenn-jocher** [comment]
Thanks for this example contribution. It's been sitting with a `CHANGES_REQUESTED` review and went stale (the inactivity notice fired on June 6, after the May 29 request to re-request review when ready), so I'm closing it to keep the queue tidy. No problem at all to reopen or re-submit when you're ready to pick it back up.

If a real API key was ever committed to this branch's history at any point, please rotate/revoke it as a precaution even though the current diff uses an environment variable. Thanks again, and feel free to ping for a re-review when it's ready.
